server/gfmd/db_ldap.c

```diff
@@ -3264,6 +3264,162 @@ gfarm_ldap_quota_dir_load(void *closure,
 	return (GFARM_ERR_FUNCTION_NOT_IMPLEMENTED);
 }
 
+/**********************************************************************/
+
+static char *gfarm_ldap_user_auth_make_dn(void *vkey);
+static gfarm_error_t gfarm_ldap_user_auth_set_field(void *info,
+	char *attribute, char **vals);
+
+struct gfarm_ldap_user_auth_key {
+	const char *username;
+	const char *auth_method;
+};
+
+static const struct gfarm_ldap_generic_info_ops gfarm_ldap_user_auth_ops = {
+	&gfarm_base_user_auth_ops,
+	"(objectclass=GFarmUserAuth)",
+	"username=%s, auth_method=%s, %s",
+	gfarm_ldap_user_info_make_dn,
+	gfarm_ldap_user_info_set_field,
+};
+
+static char *
+gfarm_ldap_user_auth_make_dn(void *vkey)
+{
+	struct gfarm_ldap_user_auth_key *key = vkey;
+	char *dn;
+
+	GFARM_MALLOC_ARRAY(dn, strlen(gfarm_ldap_user_auth_ops.dn_template) +
+		strlen(key->username) + strlen(key->auth_method) +
+		strlen(gfarm_ldap_base_dn) + 1);
+	if (dn == NULL) {
+		gflog_debug(GFARM_MSG_1002128,
+			"allocation of string 'dn' failed");
+		return (NULL);
+	}
+	sprintf(dn, gfarm_ldap_user_auth_ops.dn_template,
+		key->username, key->auth_method, gfarm_ldap_base_dn);
+	return (dn);
+}
+
+static gfarm_error_t
+gfarm_ldap_user_auth_set_field(
+	void *vinfo,
+	char *attribute,
+	char **vals)
+{
+	gfarm_error_t err = GFARM_ERR_NO_ERROR;
+	struct db_user_auth_arg *info = vinfo;
+	static const char diag[] = "gfarm_ldap_user_auth_set_field";
+
+	if (strcasecmp(attribute, "username") == 0) {
+		info->username = strdup_log(vals[0], diag);
+		if (info->username == NULL)
+			err = GFARM_ERR_NO_MEMORY;
+	} else if (strcasecmp(attribute, "authMethod") == 0) {
+		/* XXX FIXME - hack to allow null string */
```

Comment:
null string is not allowed for authMethod, thus, this hack is unnecessary.
although similar hack for AuthUserId is still necessary.